### PR TITLE
Moves to the template loader hooks and functions into lib/compat folder

### PR DESCRIPTION
**Pull request by youknowriad**
See #37141 for the reasoning here.

### CODE REVIEW

File: lib/compat/wordpress-5.9/block-template.php
Comment (youknowriad) [inline lib/compat/wordpress-5.9/block-template.php:229]: I believe that function is useless and I can't find it on Core either. So either we remove it or backport it.

Comment (youknowriad) [reply lib/compat/wordpress-5.9/block-template.php:229]: @scruffian Is this something that is needed for universal themes? Should this be backported as `render_block_template` or something in Core?

File: lib/compat/wordpress-5.9/block-template.php
Comment (youknowriad) [inline lib/compat/wordpress-5.9/block-template.php:20]: just aligning function names with Core equivalents

File: lib/compat/wordpress-5.9/block-template.php
Comment (ockham) [reply lib/compat/wordpress-5.9/block-template.php:229]: Seems to me like this has been obsolete since #28319 🤔

**Review by ockham — CHANGES_REQUESTED** [review]
Since we're moving `template-canvas.php`, I believe we also need to update this path:

```diff
diff --git a/lib/compat/wordpress-5.9/block-template.php b/lib/compat/wordpress-5.9/block-template.php
index 85b0a91b4e..4a57ece0d8 100644
--- a/lib/compat/wordpress-5.9/block-template.php
+++ b/lib/compat/wordpress-5.9/block-template.php
@@ -102,7 +102,7 @@ function gutenberg_locate_block_template( $template, $type, array $templates ) {
        add_action( 'wp_head', 'gutenberg_render_title_tag', 1 ); // ...and make it unconditional.
 
        // This file will be included instead of the theme's template file.
-       return gutenberg_dir_path() . 'lib/template-canvas.php';
+       return gutenberg_dir_path() . 'lib/compat/wordpress-5.9/template-canvas.php';
 }
 
 /**
```

**Comment by youknowriad** [comment]
@ockham good catch, I actually did that change but somehow it got lost before the commit.

**Review by ockham — APPROVED** [review]
Thanks Riad, this LGTM now! 👍 

(We might want to wait for confirmation for [this question](https://github.com/WordPress/gutenberg/pull/37149/files#r762992787) before merging.)

File: lib/compat/wordpress-5.9/block-template.php
Comment (youknowriad) [reply lib/compat/wordpress-5.9/block-template.php:229]: Indeed 👍 I was confusing this with `block_template_part` which is used in hybrid themes.

**Pull request merged**
